### ISSUE TITLE
[PROPOSAL] Improve readability and hackability of the core: tabs macro

### ISSUE THREAD

**Issue by pmario** [TiddlyWiki/TiddlyWiki5#6464]
The existing tabs-macro is 8 years old, the structure is deeply nested and therefore close to impossible to read and tweak for humans. See: https://github.com/Jermolene/TiddlyWiki5/blob/d3ea98fcef6e655e9aa71cc47bf4cfa3405758f0/core/wiki/macros/tabs.tid#L1

## Proposal step 1

 - I would like to make it a bit more readable by adding indentation 
    - and use `\whitespace trim`, which wasn't available as the macro was created
 - I'll take care that the `html` output is the same as with the existing macro

## Proposal step 2

 - I'll transform every `$variable$` to the new `<<__variable__>>` format
 - html output will be the same

## Proposal step 3

 - I would like untangle the nested structure 
   - for better readability and hackability
 - As [proposed by Saq](https://github.com/Jermolene/TiddlyWiki5/issues/6464#issuecomment-1041556662) using a `condition` field in the tiddler may be an advantage. 
 - ~~I would like to add a new `exclude` parameter at the end, that allows plugin authors to dynamically remove tabs~~
   - This should allow us to add a **per tiddler** ~~exclude~~ configuration, useful to alter the tabs in the tiddler info-area
   - **The goal** is to hide tabs without the need to modify core UI tiddlers

@Jermolene ... Is there any chance that those changes splitted to 3 separate PRs will get merged?

@flibbles @saqimtiaz ... What do you think?

**Comment by saqimtiaz** [comment]
I have no objection to #1 and #2 apart from perhaps the consideration that we need to have a documented decision that we want to do this across the core, i.e. this should not be a one off improvement but part of a wider change.

Could you please clarify exactly what you mean by an `exclude parameter` and a `per tiddler` configuration, perhaps with a **simple and concise** example use case? I don't quite follow the use case or the mechanism that you have in mind.

**Comment by pmario** [comment]
>I have no objection to https://github.com/Jermolene/TiddlyWiki5/issues/1 and https://github.com/Jermolene/TiddlyWiki5/issues/2 apart from perhaps the consideration that we need to have a documented decision that we want to do this across the core, i.e. this should not be a one off improvement but part of a wider change.

That's already ongoing. See: https://github.com/Jermolene/TiddlyWiki5/pulls/flibbles
See the **Trim saga** PRs which are the first baby-steps in that direction. @flibbles wants to use that info to be able to improve the uglify project. I would like to improve readability by adding indentation.

**Comment by saqimtiaz** [comment]
@pmario the work that Flibbles is doing is about adding the trim pragma and distinct from the need for a documented decision that we want to start indenting the wikitext in the core for better readability, and that we want to start using the alternative form for referencing variables in macros. We probably also need style guidelines for indenting the wikitext.

There are other core improvements we want to make as well, off the top of my head I can think of collapsing multiple $set and $vars widgets into $let widgets, and replacing `has[draft.of]` in filters with `is[draft]`. Each of these requires an actual decision, and perhaps a ticket tracking all of these core wikitext improvements that we want to make.

**Comment by pmario** [comment]
>Could you please clarify exactly what you mean by an `exclude parameter` and a `per tiddler` configuration, perhaps with a simple and concise example use case? I don't quite follow the use case or the mechanism that you have in mind.

I want to be able to dynamically add and remove tabs from the tiddler-info area. Eg my bundler-plugin should be able to add and remove plugin-tabs according to the functionality that is possible. 

At the moment if a plugin adds a new tab with `$:/tags/TiddlerInfo` the setting is global. I want to make that mechanism dynamic, without the need to modify the core UI.

**Comment by saqimtiaz** [comment]
> I want to be able to dynamically add and remove tabs from the tiddler-info area. Eg my bundler-plugin should be able to add and remove plugin-tabs according to the functionality that is possible.

For consistency with other parts of the core, it would probably make sense adapt the macro to support a `condition` field in the tiddlers being added as tabs.

**Comment by pmario** [comment]
>For consistency with other parts of the core, it would probably make sense adapt the macro to support a condition field in the tiddlers being added as tabs.

That would also be possible. ... You are referring to the "visibility condition" for toolbar buttons. right?

**Comment by saqimtiaz** [comment]
> That would also be possible. ... You are referring to the "visibility condition" for toolbar buttons. right?

@pmario yes. It is a good mechanism and I think we need to make use of it in other places as well (such as the view template toolbar).

**Comment by saqimtiaz** [comment]
@pmario we also use it in the export related macros:

**Comment by pmario** [comment]
>@pmario yes. It is a good mechanism and I think we need to make use of it in other places as well (such as the view template toolbar).

I did update the OP

**Comment by flibbles** [comment]
@pmario,

**About Proposal 1** (whitespace trim): I'm obviously all for that. [You might even say I'm a step ahead of you](https://github.com/Jermolene/TiddlyWiki5/pull/6270/files#diff-d5f6bdd740c3562f9cd200d03bd7c3fbfeb624d703d3fb9249d79aef33fde452R5).

**About Proposal 2** (<<\_\_variable\_\_>>): I am also all for that. Placeholders are incredibly powerful, but I believe they should only be used when they're necessary, because they are also incredibly *error prone*. They've been banes to me in both Relink and Uglify.

**About Proposal 3** (exclude): This seems neat, but before we go encoding all new features, you should know that we already have all the tools we need. I changed the filter in "$:/core/ui/TiddlerInfo" to `[all[shadows+tiddlers]tag[$:/tags/TiddlerInfo]!has[draft.of]] :filter[<..currentTiddler>!filter{!!exclude}]`. And then when I added an "exclude" field to "$:/core/ui/TiddlerInfo/Fields" with a value of `[<currentTiddler>prefix[Filter]]`, that TiddlerInfo tab was then excluded from all tiddlers with "Filter" as a prefix. I'll leave it to you guys to decide if that solution is enough, or if a dedicated macro parameter is called for.

@saqimtiaz 

> There are other core improvements we want to make as well, off the top of my head I can think of collapsing multiple $set and $vars widgets into $let widgets, and replacing has[draft.of] in filters with is[draft]. Each of these requires an actual decision, and perhaps a ticket tracking all of these core wikitext improvements that we want to make.

This is unconfirmed, but I've heard rumors that somebody recently wrote an uglifying plugin which takes care of converting $vars and $set into $let, and then collapsing $let widgets together. It also converts has[draft.of] to is[draft]. So anybody who wants these improvements can have them without going through the hassle of making PRs. However, I acknowledge that the reason for both of these changes is to improve readability, which is not what this uglifying plugin is about.

**Comment by CodaCodr** [comment]
In terms of hackability, I'd love to see tabs afford a couple of render hooks:

```
<$macrocall $name=tabs tabsList=<<filter>> before=<<my-before>> after=<<my-after>> ...  />
```

which are applied thus:

```
<div class="tc-tab-content ...">
$before$
<!-- content -->
$after$
</div>
```

Use case:

```
\define my-before()
<$if value=<<flag>> match="tab-hosted">
  <$link to... />
</$if>
\end

<$let flag="tab-hosted">
<$macrocall $name=tabs tabsList=<<filter>> before=<<my-before>> ...  />
</$let>
```

<hr>

Also...

If some of this could be baked in it would be nice -- .tc-vertical doesn't scale so well without it:

```
.tc-tab-buttons.tc-vertical {
 max-height:70vh;
 overflow-y:auto;
 scrollbar-width:thin;
 width:15em;
}
tc-tab-content.tc-vertical {
 max-width:10vw;
 overflow-x:hidden;
 width:8vw;
}
```

**Comment by flibbles** [comment]
@CodaCodr, are you sure what you're describing isn't achievable with the already-present template parameter? A template could easily prescribe "before" content and "after" content for each tab pane.

**Comment by CodaCodr** [comment]
Hmm - perhaps you're right. In fact, I'm not sure why I didn't notice that (RTFM) 😊

**Comment by pmario** [comment]
Just a note:  A PR will cause trouble with: **Trim Saga: Another batch of medium-sized files** #6270 ... So it should be merged first

**Comment by pmario** [comment]
>Hmm - perhaps you're right. In fact, I'm not sure why I didn't notice that (RTFM) 😊

@CodaCodr .. could you delete the post: https://github.com/Jermolene/TiddlyWiki5/issues/6464#issuecomment-1041915367 since it seems it's not relevant anymore.

**Comment by pmario** [comment]
Updating the core macros is an ongoing process and is done -- on demand. The OP explicitly names the tab-macro, which has been updated since.  So this issue will be closed.
